### PR TITLE
Fix/rest dial timeout does not send status callback

### PR DESCRIPTION
**Pull request by xquanluu**
#403

### CODE REVIEW

File: lib/session/call-session.js [hunk 1]
```diff
@@ -1790,6 +1792,17 @@ class CallSession extends Emitter {
...
       .catch((err) => this.logger.error(err, 'redis error'));
+
+    if (this.wakeupResolver && !this.dlg && this.isCurrentCallHigherThan(CallStatus.Failed)) {
+      // Someone is waiting for call to be final
+      this.wakeupResolver({reason: 'session ended'});
+      this.wakeupResolver = null;
+    }
+  }
+
+  isCurrentCallHigherThan(target) {
```
Comment (davehorton) [inline lib/session/call-session.js]: Wow, I don't understand what this code is doing.  Can you explain ?

File: lib/session/call-session.js [hunk 2]
```diff
@@ -796,7 +796,9 @@ class CallSession extends Emitter {
         }
       }
 
-      if (0 === this.tasks.length && this.requestor instanceof WsRequestor && !this.callGone) {
+      if (0 === this.tasks.length && ((this.requestor instanceof WsRequestor && !this.callGone) ||
+        // Tasks are all done, but dialog is not establish yet, waiting for cancel/handup the call.
+        !this.dlg)) {
```
Comment (davehorton) [inline lib/session/call-session.js:801]: Are you sure that this won't hit cases other than the outdial failure?  What about an incoming call that an app simply uses a sip:decline verb on?  In that case when the app is over there has never been a dialog.  Can you make sure that case is handled appropriately?

File: lib/session/call-session.js (hunk 1 above)
Comment (davehorton) [reply lib/session/call-session.js]: what does "higher than" mean in this case?

Comment (xquanluu) [reply lib/session/call-session.js]: For example: Call is final when callStatus is in [Faled, Busy, NoAnswer, Completed]

Here the function is has typo issue that I change name to isCurrentCallStatusHigherThan

I want to use index of each key,value pair in json to identify if call is already final

File: lib/session/call-session.js (hunk 2 above)
Comment (xquanluu) [reply lib/session/call-session.js]: you are right, add fix to outbound call only.

**Comment by xquanluu** [comment]
sure, Let me think on another approach actually

**Pull request closed without merging**